### PR TITLE
DDF-04284 Poll interval configuration for OpensearchSource

### LINKED ISSUE #4284: OpenSearchSource doesn't have a configurable poll interval
### Description

Improvement: OpenSearchSource doesn't have a configurable poll interval

We have a server admin that would like us to stop polling him every minute. Other sources allow you to change the poll interval, but not this one.

Issue is not a bug. 

### Steps to Reproduce (if issue is bug)

1. [First Step]
2. [Second Step]
3. [and so on...]

#### Expected behavior:
[What you expect to happen]

#### Actual behavior:
[What actually happens]

#### Reproduces how often:
[What percentage of the time does it reproduce?]

### Version
[Application version]

### Additional Information
Any additional information, configuration or data that might be necessary
to reproduce the issue.

### PR DESCRIPTION
**Pull request by rymach**
#### What does this PR do?
Fixes #4284 
The OpenSearchSource doesn't have a configurable poll interval. We have a server admin that would like us to stop polling him every minute. Other sources allow you to change the poll interval, but not this one. This change adds Poll interval configuration for OpensearchSource.

#### Who is reviewing it? 

@leo-sakh 
@glenhein 
#### Ask 2 committers to review/merge the PR and tag them here.

@bdeining
@troymohl
@lamhuy 
#### How should this be tested?
Add a loopback source
Turn up cxf logging to watch poll
etc.
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 1]
```diff
@@ -887,7 +944,8 @@ public String getUsername() {
 
   public void setUsername(String username) {
     this.username = username;
-    factory = null;
+    updateFactory();
+    updateScheduler();
```
Comment (rymach) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: the downside of not using component-managed properties is that every setter is called when properties are updated through the administration ui. the result is that we do a lot of wasted work creating factories and schedulers that are never really needed (only the last setter that is called matters). however, I still prefer the straightforward style of container managed properties.

**Comment by rymach** [comment]
build now

**Comment by ethantmanns** [comment]
Could we add more to the PR description?

**Comment by rymach** [comment]
> @rymach , can you update the pull request template to provide a description of what the PR does, and the commit messages to follow the Codice DDF guidelines https://github.com/codice/ddf/wiki/Git-Commit-Message-Guidelines

done

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 2]
```diff
@@ -103,6 +107,7 @@
  * Federated site that talks via OpenSearch to the DDF platform. Communication is usually performed
  * via https which requires a keystore and trust store to be provided.
  */
+@SuppressWarnings("WeakerAccess")
```
Comment (gjvera) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: ✏️ Usually when we suppress warnings we add a comment for why e.g.
`@SuppressWarnings("WeakerAccess" /* Needed by blueprint. */)` or something of that nature.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 3]
```diff
@@ -168,16 +175,23 @@
 
   protected final OpenSearchFilterVisitor openSearchFilterVisitor;
 
-  protected Integer connectionTimeout;
+  protected Integer connectionTimeout = 30000;
```
Comment (gjvera) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: ❓ Should these be configurable? Any reason you chose 30000 and 60000?

File: catalog/opensearch/catalog-opensearch-source/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 4]
```diff
@@ -71,6 +71,10 @@
             name="Receive Timeout" id="receiveTimeout"
             required="true" type="Integer" default="60000"/>
 
+        <AD description="Poll Interval to Check if the Source is available (in minutes - minimum 1)."
```
Comment (gjvera) [inline catalog/opensearch/catalog-opensearch-source/src/main/resources/OSGI-INF/metatype/metatype.xml]: It looks like connection timeout and receive timeout are supposed to be configurable and shouldn't be hardcoded.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java (hunk 3 above)
Comment (rymach) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: those are the default values. I set them b/c blueprint sets the properties onesie twosies and I create the `factory` object with around five fields. all of the weird things I don't like about this PR are things that I couldn't figure out a better way to do w/ blueprint

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java (hunk 2 above)
Comment (rymach) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: removed it

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 5]
```diff
@@ -811,6 +860,11 @@ public void setDistanceTolerance(int distanceTolerance) {
     }
   }
 
+  public void setPollInterval(Integer interval) {
+    this.pollInterval = interval;
```
Comment (willwill96) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: The metatype file says this value has a minimum value of 1. Should we be enforcing that here?

Comment (rymach) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: good catch

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 6]
```diff
@@ -861,7 +861,7 @@ public void setDistanceTolerance(int distanceTolerance) {
   }
 
   public void setPollInterval(Integer interval) {
-    this.pollInterval = interval;
+    this.pollInterval = Math.max(1, interval);
```
Comment (gjvera) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: ✏️ Not sure if you agree, but should we have a log statement here that says it's defaulting to 1 if the interval is less than 1?

**Comment by bdeining** [comment]
build now

**Comment by gjvera** [comment]
### 🎉 Hero Successful 🎉 
quickbuilt/installed
set up an opensearch source and set different poll intervals
set some negative ones and it defaulted to 1 
whenever the poll interval was updated the following was in the logs 
```
12:46:36,549 | DEBUG | chSourceThread 0 | ddf.opensearch.source.OpenSearchSource$1  263 | og-opensearch-source | Checking availability for source DDF-4284
12:46:36,558 | DEBUG | pool-22-thread-1 | e.ddf.opensearch.source.OpenSearchSource  247 | og-opensearch-source | Setting Availability poll task for 10 minute(s) on Source DDF-4284
```
set `log:set INFO org.apache.cxf` and saw the following in the logs
```
12:57:22,708 | DEBUG | chSourceThread 0 | ddf.opensearch.source.OpenSearchSource$1  263 | og-opensearch-source | Checking availability for source DDF-4284
12:57:22,718 | INFO  | chSourceThread 0 | f.interceptor.AbstractLoggingInterceptor  278 | .apache.cxf.cxf-core | Outbound Message
---------------------------
ID: 12
Address: https://localhost:8993/services/catalog/query
Http-Method: HEAD
Content-Type:
Headers: {Accept=[application/vnd.paos+xml, */*], PAOS=[ver="urn:liberty:paos:2003-08", "urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp","urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp:2.0:WantAuthnRequestsSigned"]}
--------------------------------------
12:57:22,728 | INFO  | chSourceThread 0 | f.interceptor.AbstractLoggingInterceptor  278 | .apache.cxf.cxf-core | Inbound Message
----------------------------
ID: 12
Response-Code: 200
Encoding: ISO-8859-1
Content-Type:
Headers: {Date=[Tue, 05 Mar 2019 19:57:22 GMT], Server=[Jetty(9.4.12.v20180830)], transfer-encoding=[chunked]}
--------------------------------------
```

LGTM!!!

**Comment by rymach** [comment]
build now

**Comment by rymach** [comment]
(can't be too safe)

**Comment by gjvera** [comment]
build now

**Comment by gjvera** [comment]
build now

**Comment by mojogitoverhere** [comment]
Is `testBasicFederatedAuth` a legitimate failure?

**Comment by rymach** [comment]
> Is `testBasicFederatedAuth` a legitimate failure?

looks like it. I'm looking into it now

**Comment by rymach** [comment]
build now

**Comment by rymach** [comment]
build now

**Pull request merged**